### DISCUSSION TITLE
API usage - not getting `baseTypesPath` to work

### DISCUSSION THREAD

**Discussion by JoviDeCroock** [dotansimha/graphql-code-generator#9316] [Q&A] [answered]
Hey all,

I've been using the project for [GraphQLSP](https://github.com/0no-co/GraphQLSP/pull/39), in the linked PR I have been trying to use the `typescript` plugin to generate a set of global type and then on the second run have the `typescript-operations` and `typed-document-node` import from this global file... However I have tried a lot of permutations of the `baseTypesPath` setting but it hasn't really worked out. Initially creating this as a Q&A because I'm not sure whether this is a bug or intended behavior. For clarity I am using the node API here.

**Comment by eddeee888** [comment]
Hi @JoviDeCroock 👋 

Looking through the linked PR, I think this [line](https://github.com/0no-co/GraphQLSP/pull/39/files#diff-fa6098769eb19209d3cd2122ba90397254ae7ff90034b6a10edc25148b24f0d7R73) is the right place. 

What would a typical value of `baseTypesPath` would be in this case? 
Have you tried running this through a normal Codegen config setup to see if it works? 🙂

**Reply by JoviDeCroock** [reply]
In a normal codegenerator setup it works correctly like

```yaml
    presetConfig:
      extension: .generated.ts
      baseTypesPath: ./graphql/types.ts
```

However the presetConfig thing seems unsupported in the API 😅 a typical one would be a full path i.e. /users/<project>/__generated__/x.ts

**Reply by eddeee888** [reply]
ahhh right, I see. `baseTypesPath` is mentioned in this [near-operation-file preset](https://the-guild.dev/graphql/codegen/plugins/presets/near-operation-file-preset#basetypespath) so I think it works in that preset.

So it may not work in normal `typescript-operations` or `typed-document-node` plugins. I also cannot find `baseTypesPath` in these plugin docs

Let me look into `near-operation-file preset` source code to see how it's used.

**Reply by eddeee888** [reply]
Yea I think `baseTypesPath` ends up somewhere [here in the preset](https://github.com/dotansimha/graphql-code-generator-community/blob/bf71aee287c39618c3a3f34132257ad7cacc6234/packages/presets/near-operation-file/src/resolve-document-imports.ts#L91)

I think we can't run codegen on presets. 

However, it seems like it just adds a line. I wonder if it would work if we use the `@graphql-codegen/add` plugin to add a line at the top of the generated files in your use case?

```ts
import * as addPlugin from '@graphql-codegen/add'; // 1. Import `add` plugin
// other code
const config = {
    documents: [
      {
        location: 'operation.graphql',
        document: parse(doc),
      },
    ],
    config: { /* Your other config */ },
    filename: outputFile,
    schema: parse(printSchema(schema)),
    plugins: [
      { 'typescript-operations': {} }, 
      { 'typed-document-node': {} }
      { add: { content: 'import * as Types from "..." ' } } // 2. Work out a reasonable import path here
    ],
    pluginMap: {
      'typescript-operations': typescriptOperationsPlugin,
      'typed-document-node': typedDocumentNodePlugin,
      add: addPlugin, // 3. Wire up `addPlugin`
    },
  }
```

**Reply by JoviDeCroock** [reply]
but then we'd also have to prefix all imports from typescript-operations/typed-document-node with types. right? That's some prefix option?

**Reply by eddeee888** [reply, marked as answer]
Ah great question! I think you can update your `config` to this: 

```ts
const config = {
    documents: [...],
    config: { 
      namespacedImportName: 'Types', // This adds `Types.` to your types
    },
    filename: outputFile,
    schema: parse(printSchema(schema)),
    plugins: [...],
    pluginMap: {...},
  }
}
```

`near-operation-file` preset does that [here](https://github.com/dotansimha/graphql-code-generator-community/blob/bf71aee287c39618c3a3f34132257ad7cacc6234/packages/presets/near-operation-file/src/index.ts#L328)
